### ISSUE TITLE
Jolokia stops responding within 60 seconds of process start

### ISSUE THREAD

**Issue by rbjorklin** [jolokia/jolokia#259]
Starting a java process with the JVM agent as described [here](https://jolokia.org/agent/jvm.html) I can poll for values for roughly 10-40 seconds before Jolokia stops responding. Stopping and starting the agent again from the command line as described [here](https://jolokia.org/reference/html/agents.html#jvm-attach) makes it possible to poll for another 10-40 seconds. Restarting with `--verbose` and `--debug` provides no help.

Running Jolokia 1.3.3 and java:

```
$ java -version
java version "1.8.0_92"
Java(TM) SE Runtime Environment (build 1.8.0_92-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.92-b14, mixed mode)
```

**Comment by rhuss** [comment]
* What do you mean with stop ? Does a request blocks or is the port closed ? (no connection).
* How do you start the agent, via --javaagent or by dyamically attaching it to a process ?
* When you increase the backlog for connections with the `backlog` options does this change something ?
* When you change the executor model from `single` to e.g. `cached`, does this help ? 
* Could you make thread dump when it hangs (`SIGQUIT` to the process, or CTRL-\ when process is in the foreground ?)
* What application are you instrumenting ?
* Which queries do you perform ?

**Comment by rbjorklin** [comment]
* The request blocks what seems indefinitely 
* --javaagent, restarts were made dynamically
* Changing backlog had no apparent effect
* Changing to fixed made it run for about 4-5 minutes before hanging. Cached seems to work (not tested for very long)
* For dump see [jstack.txt](https://github.com/rhuss/jolokia/files/272349/jstack.txt) jstack output is anonymized and slightly cleaned.
* In-house developed
* I've only asked for 1 value at a time. http://172.17.0.2:35518/jolokia/read/java.lang:type=Memory/HeapMemoryUsage/used

**Comment by rhuss** [comment]
Hmm, the relevant thread for Jolokia is

```
"Thread-49" #129 daemon prio=5 os_prio=0 tid=0x00007f0d6c008000 nid=0x25ea runnable [0x00007f0dae8e8000]
   java.lang.Thread.State: RUNNABLE
	at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
	at sun.nio.ch.IOUtil.read(IOUtil.java:197)
	at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380)
	- locked <0x0000000744e555d0> (a java.lang.Object)
	at sun.net.httpserver.Request$ReadStream.read(Request.java:285)
	- locked <0x000000075bd4b298> (a sun.net.httpserver.Request$ReadStream)
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:265)
	- locked <0x000000075bd4b270> (a java.io.BufferedInputStream)
	at sun.net.httpserver.Request.readLine(Request.java:80)
	at sun.net.httpserver.Request.<init>(Request.java:50)
	at sun.net.httpserver.ServerImpl$Exchange.run(ServerImpl.java:536)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

where it looks like that the agent still tries to read the request but is not finished by the client ? How do you query the agent ? Could you try it with a curl loop, too ?

**Comment by rbjorklin** [comment]
I've actually been doing a my testing with a curl loop just like this:

    while true ; do curl http://172.17.0.2:35518/jolokia/read/java.lang:type=Memory/HeapMemoryUsage/used && sleep 2 ; done

**Comment by rhuss** [comment]
Looks like you are running your app in a Docker container. Could you also try to run your loop from within the container ? E.g. do a `docker exec` into it and the query the local port there ? Behave it the same ?

**Comment by rbjorklin** [comment]
Yes the application is running within a Docker container. I've tried with the loop running both inside and outside and the behavior is the same.

**Comment by rbjorklin** [comment]
It seems the problem can be avoided by binding to the loopback interface instead of the docker default interface eth0. Unfortunately this makes it unusable since the values cannot be read from outside the container.

**Comment by rbjorklin** [comment]
It seems it's not tied to the interface but the port. If I bind to eth0 on a port that has not been exposed outside of the container it works as expected but if I change the port it binds to to one that's exposed on the host machine it stops working. This feels like an issue with Docker so I will open a ticket there as well.

**Comment by rhuss** [comment]
Thats really strange than I never had or heard of this symptoms. We use the agent seamlessly within docker containers (even OpenShift uses it), we are binding to 0.0.0.0, though.

Just in case could you try setting the option `allowDnsReverseLookup` to `false` (and set `discoveryEnabled` to `false`) to minimize other impact.

In the meantime I hope I can reproduce it (you are using an Oracle JDK, right ? On which linux distribution (e.g. which base Docker image) ?)

**Comment by rbjorklin** [comment]
Binding to 0.0.0.0 made no difference neither did setting `allowDnsReverseLookup` and `discoveryEnabled` to `false`.

Yes I'm using the Oracle JDK and running Fedora 23 on the host OS. The container is [centos:7](https://hub.docker.com/_/centos/) with the described modifications for running systemd inside it.

**Comment by rbjorklin** [comment]
Just chiming in to say that after a reboot the problem seems to be gone. The only real difference is that the kernel got bumped from 4.4.6 to 4.4.9. Feel free to close this issue.

**Comment by sabyasachisabs** [comment]
Happens with me too, everytime I have to restart the docker service.

**Comment by pauloricardomg** [comment]
I was facing a similar issue and after investigation I determined this was due to the cleanup thread wrongly stopping the jolokia HTTP server. This should be fixed by https://github.com/jolokia/jolokia/commit/82d96c04bd95cac324acc9180068b96a7e93e2ec on 2.0.3.

**Comment by rbjorklin** [comment]
Throwback to what at this point feels like another life. I’m going to take your word on this being fixed in 2.0.3 and close this.
